File: src/models/Announcement.js
const mongoose = require('mongoose');

const announcementSchema = new mongoose.Schema({
  title: {
    type: String,
    required: [true, 'Title is required'],
    trim: true,
    maxlength: [200, 'Title cannot exceed 200 characters']
  },
  content: {
    type: String,
    required: [true, 'Content is required'],
    maxlength: [2000, 'Content cannot exceed 2000 characters']
  },
  type: {
    type: String,
    enum: ['general', 'urgent', 'event', 'maintenance', 'fee', 'academic', 'holiday'],
    default: 'general'
  },
  priority: {
    type: String,
    enum: ['low', 'medium', 'high', 'critical'],
    default: 'medium'
  },
  targetAudience: {
    type: String,
    enum: ['all', 'students', 'wardens', 'admins', 'specific_rooms', 'specific_users'],
    default: 'all'
  },
  targetRooms: [{
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Room'
  }],
  targetUsers: [{
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User'
  }],
  createdBy: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: [true, 'Creator is required']
  },
  publishDate: {
    type: Date,
    default: Date.now
  },
  expiryDate: Date,
  status: {
    type: String,
    enum: ['draft', 'published', 'archived', 'expired'],
    default: 'draft'
  },
  attachments: [{
    fileName: String,
    filePath: String,
    fileSize: Number,
    uploadDate: {
      type: Date,
      default: Date.now
    }
  }],
  readBy: [{
    user: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User'
    },
    readAt: {
      type: Date,
      default: Date.now
    }
  }],
  likes: [{
    user: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User'
    },
    likedAt: {
      type: Date,
      default: Date.now
    }
  }],
  comments: [{
    user: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User'
    },
    comment: {
      type: String,
      maxlength: [500, 'Comment cannot exceed 500 characters']
    },
    commentedAt: {
      type: Date,
      default: Date.now
    }
  }],
  tags: [String],
  isSticky: {
    type: Boolean,
    default: false
  },
  emailSent: {
    type: Boolean,
    default: false
  },
  smsSent: {
    type: Boolean,
    default: false
  },
  notificationSent: {
    type: Boolean,
    default: false
  },
  viewCount: {
    type: Number,
    default: 0
  }
}, {
  timestamps: true
});

// Indexes for better query performance
announcementSchema.index({ createdBy: 1 });
announcementSchema.index({ status: 1 });
announcementSchema.index({ type: 1 });
announcementSchema.index({ priority: 1 });
announcementSchema.index({ publishDate: -1 });
announcementSchema.index({ expiryDate: 1 });
announcementSchema.index({ targetAudience: 1 });

// Pre-save middleware to update status based on expiry
announcementSchema.pre('save', function(next) {
  if (this.expiryDate && new Date() > this.expiryDate && this.status === 'published') {
    this.status = 'expired';
  }
  next();
});

// Method to mark as read by user
announcementSchema.methods.markAsRead = function(userId) {
  const existingRead = this.readBy.find(read => 
    read.user.toString() === userId.toString()
  );
  
  if (!existingRead) {
    this.readBy.push({
      user: userId,
      readAt: new Date()
    });
    this.viewCount += 1;
  }
  
  return this.save();
};

// Method to add like
announcementSchema.methods.addLike = function(userId) {
  const existingLike = this.likes.find(like => 
    like.user.toString() === userId.toString()
  );
  
  if (!existingLike) {
    this.likes.push({
      user: userId,
      likedAt: new Date()
    });
  }
  
  return this.save();
};

// Method to remove like
announcementSchema.methods.removeLike = function(userId) {
  this.likes = this.likes.filter(like => 
    like.user.toString() !== userId.toString()
  );
  
  return this.save();
};

// Method to add comment
announcementSchema.methods.addComment = function(userId, commentText) {
  this.comments.push({
    user: userId,
    comment: commentText,
    commentedAt: new Date()
  });
  
  return this.save();
};

// Virtual for read percentage
announcementSchema.virtual('readPercentage').get(function() {
  // This would need to be calculated based on total target users
  // For now, returning a simple calculation
  return this.readBy.length;
});

// Virtual for like count
announcementSchema.virtual('likeCount').get(function() {
  return this.likes.length;
});

// Virtual for comment count
announcementSchema.virtual('commentCount').get(function() {
  return this.comments.length;
});

// Virtual for active status
announcementSchema.virtual('isActive').get(function() {
  const now = new Date();
  return this.status === 'published' && 
         (!this.expiryDate || now <= this.expiryDate);
});

module.exports = mongoose.model('Announcement', announcementSchema);

File: src/main/java/com/hostel/model/Announcement.java
package com.hostel.model;

import org.springframework.data.annotation.Id;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.mongodb.core.mapping.Document;

import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.Size;
import java.time.LocalDateTime;
import java.util.List;
import java.util.ArrayList;

@Document(collection = "announcements")
public class Announcement {
    @Id
    private String id;
    
    @NotBlank(message = "Title is required")
    @Size(max = 200, message = "Title cannot exceed 200 characters")
    private String title;
    
    @NotBlank(message = "Content is required")
    @Size(max = 2000, message = "Content cannot exceed 2000 characters")
    private String content;
    
    private String type = "GENERAL"; // GENERAL, URGENT, EVENT, MAINTENANCE, FEE, ACADEMIC, HOLIDAY
    
    private String priority = "MEDIUM"; // LOW, MEDIUM, HIGH, CRITICAL
    
    private String targetAudience = "ALL"; // ALL, STUDENTS, WARDENS, ADMINS, SPECIFIC_ROOMS, SPECIFIC_USERS
    
    private List<String> targetRooms = new ArrayList<>();
    
    private List<String> targetUsers = new ArrayList<>();
    
    private String createdBy;
    
    private LocalDateTime publishDate;
    
    private LocalDateTime expiryDate;
    
    private String status = "DRAFT"; // DRAFT, PUBLISHED, ARCHIVED, EXPIRED
    
    private List<Attachment> attachments = new ArrayList<>();
    
    private List<ReadBy> readBy = new ArrayList<>();
    
    private List<Like> likes = new ArrayList<>();
    
    private List<Comment> comments = new ArrayList<>();
    
    private List<String> tags = new ArrayList<>();
    
    private boolean isSticky = false;
    
    private boolean emailSent = false;
    
    private boolean smsSent = false;
    
    private boolean notificationSent = false;
    
    private Integer viewCount = 0;
    
    @CreatedDate
    private LocalDateTime createdAt;
    
    @LastModifiedDate
    private LocalDateTime updatedAt;
    
    // Constructors
    public Announcement() {}
    
    public Announcement(String title, String content, String createdBy) {
        this.title = title;
        this.content = content;
        this.createdBy = createdBy;
        this.publishDate = LocalDateTime.now();
    }
    
    // Getters and Setters
    public String getId() { return id; }
    public void setId(String id) { this.id = id; }
    
    public String getTitle() { return title; }
    public void setTitle(String title) { this.title = title; }
    
    public String getContent() { return content; }
    public void setContent(String content) { this.content = content; }
    
    public String getType() { return type; }
    public void setType(String type) { this.type = type; }
    
    public String getPriority() { return priority; }
    public void setPriority(String priority) { this.priority = priority; }
    
    public String getTargetAudience() { return targetAudience; }
    public void setTargetAudience(String targetAudience) { this.targetAudience = targetAudience; }
    
    public List<String> getTargetRooms() { return targetRooms; }
    public void setTargetRooms(List<String> targetRooms) { this.targetRooms = targetRooms; }
    
    public List<String> getTargetUsers() { return targetUsers; }
    public void setTargetUsers(List<String> targetUsers) { this.targetUsers = targetUsers; }
    
    public String getCreatedBy() { return createdBy; }
    public void setCreatedBy(String createdBy) { this.createdBy = createdBy; }
    
    public LocalDateTime getPublishDate() { return publishDate; }
    public void setPublishDate(LocalDateTime publishDate) { this.publishDate = publishDate; }
    
    public LocalDateTime getExpiryDate() { return expiryDate; }
    public void setExpiryDate(LocalDateTime expiryDate) { this.expiryDate = expiryDate; }
    
    public String getStatus() { return status; }
    public void setStatus(String status) { this.status = status; }
    
    public List<Attachment> getAttachments() { return attachments; }
    public void setAttachments(List<Attachment> attachments) { this.attachments = attachments; }
    
    public List<ReadBy> getReadBy() { return readBy; }
    public void setReadBy(List<ReadBy> readBy) { this.readBy = readBy; }
    
    public List<Like> getLikes() { return likes; }
    public void setLikes(List<Like> likes) { this.likes = likes; }
    
    public List<Comment> getComments() { return comments; }
    public void setComments(List<Comment> comments) { this.comments = comments; }
    
    public List<String> getTags() { return tags; }
    public void setTags(List<String> tags) { this.tags = tags; }
    
    public boolean isSticky() { return isSticky; }
    public void setSticky(boolean sticky) { isSticky = sticky; }
    
    public boolean isEmailSent() { return emailSent; }
    public void setEmailSent(boolean emailSent) { this.emailSent = emailSent; }
    
    public boolean isSmsSent() { return smsSent; }
    public void setSmsSent(boolean smsSent) { this.smsSent = smsSent; }
    
    public boolean isNotificationSent() { return notificationSent; }
    public void setNotificationSent(boolean notificationSent) { this.notificationSent = notificationSent; }
    
    public Integer getViewCount() { return viewCount; }
    public void setViewCount(Integer viewCount) { this.viewCount = viewCount; }
    
    public LocalDateTime getCreatedAt() { return createdAt; }
    public void setCreatedAt(LocalDateTime createdAt) { this.createdAt = createdAt; }
    
    public LocalDateTime getUpdatedAt() { return updatedAt; }
    public void setUpdatedAt(LocalDateTime updatedAt) { this.updatedAt = updatedAt; }
    
    public int getLikeCount() {
        return likes != null ? likes.size() : 0;
    }
    
    public int getCommentCount() {
        return comments != null ? comments.size() : 0;
    }
    
    public boolean isActive() {
        LocalDateTime now = LocalDateTime.now();
        return "PUBLISHED".equals(status) && 
               (expiryDate == null || now.isBefore(expiryDate));
    }
    
    // Nested classes
    public static class Attachment {
        private String fileName;
        private String filePath;
        private Long fileSize;
        private LocalDateTime uploadDate;
        
        // Getters and Setters
        public String getFileName() { return fileName; }
        public void setFileName(String fileName) { this.fileName = fileName; }
        
        public String getFilePath() { return filePath; }
        public void setFilePath(String filePath) { this.filePath = filePath; }
        
        public Long getFileSize() { return fileSize; }
        public void setFileSize(Long fileSize) { this.fileSize = fileSize; }
        
        public LocalDateTime getUploadDate() { return uploadDate; }
        public void setUploadDate(LocalDateTime uploadDate) { this.uploadDate = uploadDate; }
    }
    
    public static class ReadBy {
        private String userId;
        private LocalDateTime readAt;
        
        public ReadBy() {}
        
        public ReadBy(String userId) {
            this.userId = userId;
            this.readAt = LocalDateTime.now();
        }
        
        // Getters and Setters
        public String getUserId() { return userId; }
        public void setUserId(String userId) { this.userId = userId; }
        
        public LocalDateTime getReadAt() { return readAt; }
        public void setReadAt(LocalDateTime readAt) { this.readAt = readAt; }
    }
    
    public static class Like {
        private String userId;
        private LocalDateTime likedAt;
        
        public Like() {}
        
        public Like(String userId) {
            this.userId = userId;
            this.likedAt = LocalDateTime.now();
        }
        
        // Getters and Setters
        public String getUserId() { return userId; }
        public void setUserId(String userId) { this.userId = userId; }
        
        public LocalDateTime getLikedAt() { return likedAt; }
        public void setLikedAt(LocalDateTime likedAt) { this.likedAt = likedAt; }
    }
    
    public static class Comment {
        private String userId;
        private String comment;
        private LocalDateTime commentedAt;
        
        public Comment() {}
        
        public Comment(String userId, String comment) {
            this.userId = userId;
            this.comment = comment;
            this.commentedAt = LocalDateTime.now();
        }
        
        // Getters and Setters
        public String getUserId() { return userId; }
        public void setUserId(String userId) { this.userId = userId; }
        
        public String getComment() { return comment; }
        public void setComment(String comment) { this.comment = comment; }
        
        public LocalDateTime getCommentedAt() { return commentedAt; }
        public void setCommentedAt(LocalDateTime commentedAt) { this.commentedAt = commentedAt; }
    }
}